src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Utf8JsonReader.cs

```diff
@@ -0,0 +1,323 @@
+        {
+            if (returnType == null)
+                throw new ArgumentNullException(nameof(returnType));
+
+            return ReadValueCore(ref reader, returnType, options);
+        }
+
+        private static object ReadValueCore(ref Utf8JsonReader reader, Type returnType, JsonSerializerOptions options)
+        {
+            if (options == null)
+            {
+                options = JsonSerializerOptions.s_defaultOptions;
+            }
+
+            ReadStack readStack = default;
+            readStack.Current.Initialize(returnType, options);
+
+            ReadValueCore(options, ref reader, ref readStack);
+
+            return readStack.Current.ReturnValue;
+        }
+
+        private static void CheckSupportedOptions(JsonReaderOptions readerOptions, string paramName)
+        {
+            if (readerOptions.CommentHandling == JsonCommentHandling.Allow)
+            {
+                throw new ArgumentException(SR.JsonSerializerDoesNotSupportComments, paramName);
+            }
+        }
+
+        private static void ReadValueCore(JsonSerializerOptions options, ref Utf8JsonReader reader, ref ReadStack readStack)
+        {
+            JsonReaderState state = reader.CurrentState;
+            CheckSupportedOptions(state.Options, nameof(reader));
+
+            // Value copy to overwrite the ref on an exception and undo the destructive reads.
+            Utf8JsonReader restore = reader;
+
+            ReadOnlySpan<byte> valueSpan = default;
+            ReadOnlySequence<byte> valueSequence = default;
+
+            try
+            {
+                switch (reader.TokenType)
+                {
+                    // A new reader was created and has never been read,
+                    // so we need to move to the first token.
+                    // (or a reader has terminated and we're about to throw)
+                    case JsonTokenType.None:
+                    // Using a reader loop the caller has identified a property they wish to
+                    // hydrate into a JsonDocument. Move to the value first.
+                    case JsonTokenType.PropertyName:
+                        {
+                            if (!reader.Read())
+                            {
+                                ThrowHelper.ThrowJsonReaderException(ref reader, ExceptionResource.ExpectedOneCompleteToken);
+                            }
+                            break;
+                        }
+                }
+
+                switch (reader.TokenType)
+                {
+                    // Any of the "value start" states are acceptable.
+                    case JsonTokenType.StartObject:
+                    case JsonTokenType.StartArray:
+                        {
+                            long startingOffset = reader.TokenStartIndex;
+
+                            if (!reader.TrySkip())
```

Comment:
It would be more performant and much less code if we just call into the normal serializer loop here instead of pre-reading. We have some work to do however with checking the depth on array\object end token (in order to return at the correct time) since that code was removed recently.

Reply:
I was weary about injecting code complexity into that loop, which affects all scenarios. Since this is about reading a single value (and not some token within an arbitrary state within the payload), it was cleaner to do so up front. I measured and it should improve the ASPNET's scenario ~2x (instead of the potential ~2.5x). I don't yet see modifying the existing loop as a clear win.